include: sys: util: implement `NUM_VA_ARGS`

ycsin opened the pull request:
This macro count the number of arguments in the variable arguments list.

Added test for it in `tests/lib/sys_util`.

[Doc preview](https://builds.zephyrproject.io/zephyr/pr/71890/docs/kernel/util/index.html#c.NUM_VA_ARGS_LESS_1)

ycsin committed:
include: sys: util: implement `NUM_VA_ARGS`

This macro count the number of arguments in the variable
arguments list.

Added test for it in `tests/lib/sys_util`.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

pdgendt reviewed (dismissed):
Nice

de-nordic commented on include/zephyr/sys/util_internal.h:
```diff
@@ -136,6 +136,15 @@
+#define NUM_VA_ARGS_IMPL(...)                                                                      \
+	COND_CODE_1(                                                                               \
+		IS_EMPTY(__VA_ARGS__), (0),                                                        \
+		(Z_GET_ARG_64(__VA_ARGS__, 63, 62, 61, 60, 59, 58, 57, 56, 55, 54, 53, 52, 51, 50, \
```
Question: Couldn't this part be replaced with `UTIL_INC(NUM_VA_ARGS_LESS_1(__VA_ARGS__))`?

ycsin replied:
Didn't try this myself, but the lowest value from `NUM_VA_ARGS_LESS_1` + `UTIL_INC` should be `1` whereas `NUM_VA_ARGS` is `0`

pdgendt replied:
The following appears to work:

```c
/* Implementation details for NUM_VA_ARGS */
#define NUM_VA_ARGS_IMPL(...)                                                                      \
	COND_CODE_1(                                                                               \
		IS_EMPTY(__VA_ARGS__), (0),                                                        \
		(UTIL_INC(NUM_VA_ARGS_LESS_1(__VA_ARGS__))))
```

ycsin replied:
@pdgendt thanks for the suggestion, applied the change.
@de-nordic Sorry that I misunderstood your suggestion previously

ycsin committed:
include: sys: util: add note and test for NUM_VA_ARGS_LESS_1

Notes that `NUM_VA_ARGS_LESS_1` support up to 64 arguments,
and added test for it in `tests/lib/sys_util`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
- Did the implementation inplace in `include/zephyr/sys/util_macro.h` instead of `include/zephyr/sys/util_internal.h`, and
- Added test for `NUM_VA_ARGS_LESS_1`

@pdgendt @peter-mitsis please take another look, thanks.
